Show permitted endpoints on token download page

EdwardAndress opened the pull request:
- Grab permissions from the Token object, rather than the JWT, since the latter is only created on download
- Show permissions to user so that they can easily see what they can access (and whether this meets their requirements)
- This also allows the user to verify that the regex are valid!  At the moment, we rely on the admin users not making any errors there (there is no validation).
- Also refactor ProvisionToken service object to split on any newline char (`#split`) defaults to split on either `\n` or `\r\n`

EdwardAndress committed:
Show permitted endpoints on token download page

EdwardAndress committed:
Split on \n (Mac) or \r\n (Win) in ProvisionToken

EdwardAndress committed:
ENV['TEAM_EMAIL'] is obfuscated on Travis

himal-mandalia commented on app/controllers/tokens_controller.rb:
```diff
@@ -28,4 +29,8 @@ def set_token
     ENV['TEAM_EMAIL'] || 'nomisapi@digital.justice.gov.uk'
   end
+
+  def set_parsed_permissions
```
This could be a `before_action` filter called before the #new action.

EdwardAndress replied:
I considered that but, since it's only being used on one action (hence, no repetition to DRY out) and it's a non-standard instance var, I felt that it was better to be more explicit and set it inside the controller.

If you still think it's better to use a `before_action` filter I am happy to change it.

himal-mandalia replied:
No that's fine - it is only a one off use case so can leave it as it is.

himal-mandalia commented on app/controllers/tokens_controller.rb:
```diff
@@ -28,4 +29,8 @@ def set_token
   end
+
+  def set_parsed_permissions
+    @token.permissions.split.map { |p| Regexp.new(p).inspect }
```
This causes an error. `eval(@token.permissions)` instead works fine.

<img width="1050" alt="screen shot 2017-10-18 at 15 27 35" src="https://user-images.githubusercontent.com/4628936/31724172-ef876d4c-b418-11e7-8691-ca7e1da6feee.png">

EdwardAndress replied:
is it possible the permissions are malformed there?

EdwardAndress replied:
i.e. is the square bracket in ...`char-class: /["`... supposed to be there?

EdwardAndress replied:
you might need to reset your DB

himal-mandalia replied:
This was a new Token record I created to test with.

himal-mandalia reviewed (commented):
Few issues, see inline comments.

himal-mandalia reviewed (commented):
Need to fix up an issue in the `TokensController#new` where the a token with a nil `trackback_token` will be fetched.

Suggested fix:

```
@token = Token.find_by(trackback_token: params[:trackback_token])
    raise ActionController::RoutingError.new('Not Found') if params[:trackback_token].blank? || @token.nil?
```

EdwardAndress committed:
Raise error if params[:trackback_token].blank?

- Tokens auto-created for our envs have no trackback_token
- Those tokens can be recalled with TokensController#set_token in error

EdwardAndress commented:
@himal-mandalia FIXED and tested

himal-mandalia reviewed (approved):
Looks good

EdwardAndress commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Confirm which endpoints a token grants access to, in the email that contains the download link](https://trello.com/c/1Z1Q4gI1/469-confirm-which-endpoints-a-token-grants-access-to-in-the-email-that-contains-the-download-link)
